ibis/backends/sql/compilers/sqlite.py

```diff
@@ -334,17 +331,46 @@ def visit_TimestampTruncate(self, op, *, arg, unit):
 
     def visit_DateArithmetic(self, op, *, left, right):
         unit = op.right.dtype.unit
-        sign = "+" if isinstance(op, ops.DateAdd) else "-"
-        if unit not in (IntervalUnit.YEAR, IntervalUnit.MONTH, IntervalUnit.DAY):
+        if unit in (
+            IntervalUnit.WEEK,
+            IntervalUnit.MICROSECOND,
+            IntervalUnit.NANOSECOND,
+        ):
             raise com.UnsupportedOperationError(
-                "SQLite does not allow binary op {sign!r} with INTERVAL offset {unit}"
+                f"SQLite does not support `{unit}` units in temporal arithmetic"
             )
-        if isinstance(op.right, ops.Literal):
-            return self.f.date(left, f"{sign}{op.right.value} {unit.plural}")
+
+        right = right.this
+        if unit == IntervalUnit.MILLISECOND:
+            floor = False
+            unit = IntervalUnit.SECOND
+            right /= 1e3
         else:
-            return self.f.date(left, self.f.concat(sign, right, f" {unit.plural}"))
+            floor = True
+
+        sign = "+" if isinstance(op, (ops.DateAdd, ops.TimestampAdd)) else "-"
+
+        args = [
+            left,
+            self.f.concat(
+                sign, self.cast(right, dt.string), " ", unit.singular.lower()
+            ),
+        ]
+
+        if floor:
+            args.append("floor")
 
-    visit_DateAdd = visit_DateSub = visit_DateArithmetic
+        if isinstance(op, (ops.TimestampAdd, ops.TimestampSub)):
+            func = self.f.datetime
+            args.append("subsec")
+        else:
+            func = self.f.date
```

Comment:
This entire method became the absolute worst.

Reply:
Yep, that's pretty awful.  But hey, at least it's adding support for 4 different targets!